IanGRap/Squish
# Hats!

Contributors: Austin, Wesley

Goals: 
- Introduce an expandable system for selecting hats and spawning them on players once in the match
- Model the new hat system after the skin system so that they are similar to modify / expand

What Changed:
- ControllerSignIn.cs
    - Removed the ability to change skins with Y
    - Added ability to cycle through skins in 2 directions with the Left and Right DPad buttons
    - Added in listeners for Up and Down DPad buttons to cycle through hats
    - Added in very simple keyboard controls so that Wesley can see what hats look like in Unity (H and J to cycle on keyboard)
- Hat.cs
    - A class to store Hat data and a function for putting a hat on a player
    - This class is very similar in spirit to Skin.cs
- PlayerHeadLocation.cs
    - This script literally just exists so that I can get the location of the top of Skelegoo's head through scripts in a clean way.
    - This script allows me to call `player.GetComponent<PlayerHeadLocation>()` to get that particular spot.
- PlayerSpawner.cs
    - Added in a call to SetStartingHat() when a player first spawns in
- ActivePlayers.cs
    - Added in a variable to store the hats that players chose in the lobby so that it is remembered throughout the game
-  PlayerColorController.cs --> PlayerCosmeticSetup.cs
    - I renamed PlayerColorController.cs to PlayerCosmeticSetup.cs 
    - This was the file that set the players' skins when they are first spawned into the match scene
    - I added functionality to this file so that it also spawns in the player hats, so I figured a rename would be appropriate
- UIController.cs
    - A race condition error appeared (once again), and I noticed that the Initialize() function was moved from Awake to Start, which was causing the SceneCreator to crash. I moved it back into Awake and renamed Initialize() to OnAwake() to avoid any further confusion.
- Temporary tutorial UI was added to the  lobby
- I added in a basic visual effect for changing hats. Subject to change in the future.
- Dummy Player Prefab:
    - I had to remake the Dummy Player prefab so that the win screen could display the hats
    - The only difference is the new prefab has the PlayerHeadLocation script on it.
- I had to alter the WinScreenController object for this PR, since I had to change out the Dummy Player prefab for the new one.
    - Hopefully this doesn't cause any nasty merge conflicts

How to Test:
- Hop into the lobby and make sure you can:
    - No longer change skins with Y
    - Cycle through skins with left and right dpad
    - Cycle through hats with up and down dpad 
        - Unlike skins, hats are not unique to one player. This means 2 players can have the same hat. This was a design deceision on my end, open to change if need be.
    - Spawn in and observe that the default hat is no hat.
    - Dab and ensure that the hat follows the skull
- Start a game and ensure that all players were spawned in with their chosen hats
- Win a game and ensure that all players were spawned in with their chosen hats (in win screen)
- Test all of the above in a build
- Start the match scene without going through the lobby and ensure that it is still playable.
    - Players should all spawn in with no hat.

NOTE: 
- This PR introduced a new problem where some of the hats block the text of the win screen.
- I feel that it is out of the scope of this feature, but we will need to at least slightly redesign the win screen to compensate for this.
